src/TreeSupport.cpp

```diff
@@ -265,58 +265,71 @@ void TreeSupport::dropNodes(const SliceDataStorage& storage, std::vector<std::un
-                                difference = normal(difference, maximum_move_distance);
-                            }
-                            next_position = node.position + difference;
+                            difference = normal(difference, maximum_move_distance);
                         }
-
-                        const bool to_buildplate = !model_avoidance[branch_radius_sample][layer_nr - 1].inside(next_position);
-                        Node next_node(next_position, node.distance_to_top + 1, node.skin_direction, node.support_roof_layers_below - 1, to_buildplate);
-                        insertDroppedNode(contact_nodes[layer_nr - 1], next_node); //Insert the node, resolving conflicts of the two colliding nodes.
+                        next_position = node.position + difference;
                     }
-                    else
+
+                    const bool to_buildplate = !model_avoidance[branch_radius_sample][layer_nr - 1].inside(next_position);
+                    Node next_node(next_position, node.distance_to_top + 1, node.skin_direction, node.support_roof_layers_below - 1, to_buildplate);
+                    insertDroppedNode(contact_nodes[layer_nr - 1], next_node); //Insert the node, resolving conflicts of the two colliding nodes.
+                }
+                else if (neighbours.size() > 1) //Don't merge leaf nodes because we would then incur movement greater than the maximum move distance.
+                {
+                    //Remove all neighbours that are too close and merge them into this node.
+                    for (const Point& neighbour : neighbours)
                     {
-                        //We'll drop this node, but modify some of the properties of its neighbour.
-                        Node& neighbour = nodes_per_part[group_index][neighbours[0]];
-                        neighbour.distance_to_top = std::max(neighbour.distance_to_top, node.distance_to_top);
-                        neighbour.support_roof_layers_below = std::max(neighbour.support_roof_layers_below, node.support_roof_layers_below);
+                        if (vSize2(neighbour - node.position) < maximum_move_distance * maximum_move_distance)
+                        {
+                            const Node& neighbour_node = nodes_per_part[group_index][neighbour];
+                            node.distance_to_top = std::max(node.distance_to_top, neighbour_node.distance_to_top);
+                            node.support_roof_layers_below = std::max(node.support_roof_layers_below, neighbour_node.support_roof_layers_below);
+                            to_delete.insert(neighbour_node);
+                        }
                     }
                 }
             }
             //In the second pass, move all middle nodes.
             for (std::pair<Point, Node> entry : nodes_per_part[group_index])
             {
                 Node node = entry.second;
+                if (to_delete.find(node) != to_delete.end())
```

Comment:
Why not delete everything from to_delete from nodes_per_part before entering this loop? Seems (a bit) cleaner to me.

Reply:
We can't modify `nodes_per_part` when we want to delete the nodes because we are at that point iterating over that set. It could be solved by making a third loop in between that goes over all nodes and copies them to a different set if they don't need to be deleted and then modifying the last loop to iterate over the new set, but then I think this is cleaner (and faster).